client/app/components/onboardingGuide/OnboardingGuide.jsx

```diff
@@ -0,0 +1,203 @@
+    }).isRequired,
+    railsContext: PropTypes.object.isRequired,
+    data: PropTypes.shape({
+      path: PropTypes.string.isRequired,
+      pathHistoryForward: PropTypes.bool,
+      name: PropTypes.string.isRequired,
+      translations: PropTypes.object.isRequired,
+      onboarding_data: PropTypes.objectOf(PropTypes.shape({
+        info_image: PropTypes.string.isRequired,
+        link: PropTypes.string.isRequired,
+        complete: PropTypes.bool.isRequired,
+      }).isRequired).isRequired,
+    }).isRequired,
+  };
+
+  componentDidMount() {
+    window.addEventListener('popstate', this.handlePopstate);
+  }
+
+  componentWillUnmount() {
+    window.removeEventListener('popstate', this.handlePopstate);
+  }
+
+  componentWillUpdate(nextProps, nextState) {
+    // Back button clicks should not be saved with history.pushState
+    if (nextProps.data.pathHistoryForward) {
+      const path = nextProps.data.path;
+      this.setPushState({ path }, path, path);
+    }
+  }
+
+  constructor(props, context) {
+    super(props, context);
+
+    this.setPushState = this.setPushState.bind(this);
+    this.handlePopstate = this.handlePopstate.bind(this);
+    this.handlePageChange = this.handlePageChange.bind(this);
+
+    const paths = getPaths(props, 'getting_started_guide');
+    this.initialPath = paths.initialPath;
+    this.componentSubPath = paths.componentSubPath;
+
+    // Figure out the next step. I.e. what is the action we recommend for admins
+    this.nextStep = nextStep(this.props.data.onboarding_data,
+      translate(this.props.data.translations.next_step),
+      this.initialPath);
+
+    // Add current path to window.history. Initially it contains null as a state
+    this.setPushState(
+      { path: this.componentSubPath },
+      this.componentSubPath,
+      this.componentSubPath);
+  }
+
+  setPushState(state, title, path) {
+    // React has an internal variable 'canUseDOM', which we emulate here.
+    const canUseDOM = !!(typeof window !== 'undefined' &&
+                          window.document &&
+                          window.document.createElement);
+    const canUsePushState = !!(typeof history !== 'undefined' &&
+                                history.pushState);
+
+    if (canUseDOM && canUsePushState) {
+      window.history.pushState(state, title, `${this.initialPath}${path}`);
+    }
+  }
+
+  handlePopstate(event) {
+    if (event.state != null && event.state.path != null) {
+      this.props.actions.updateGuidePage(event.state.path, false);
+    } else if (event.state == null && typeof this.props.data.pathHistoryForward !== 'undefined') {
+      // null state means that page component's root path is reached and
+      // previous page is actually on Rails side - i.e. one step further
+      // Safari fix: if pathHistoryForward is not defined, its initial page load
+      window.history.back();
+    }
+  }
+
+  handlePageChange(path) {
+    this.props.actions.updateGuidePage(path, true);
```

Comment:
Let's leave this decision to next project. (We need to think about routing a bit more later.)
Currently, flow goes so that: click to link calls this action -> reducer -> redux store -> props changed -> this current *root component* handles window.history related actions.